pkg/volume/csi/csi_mounter.go

```diff
@@ -197,31 +197,31 @@ func (c *csiMountMgr) SetUpAt(dir string, fsGroup *int64) error {
 	}
 
 	// apply volume ownership
-	if !c.readOnly && fsGroup != nil {
-		err := volume.SetVolumeOwnership(c, fsGroup)
-		if err != nil {
-			// attempt to rollback mount.
-			glog.Error(log("mounter.SetupAt failed to set fsgroup volume ownership for [%s]: %v", c.volumeID, err))
-			glog.V(4).Info(log("mounter.SetupAt attempting to unpublish volume %s due to previous error", c.volumeID))
-			if unpubErr := csi.NodeUnpublishVolume(ctx, c.volumeID, dir); unpubErr != nil {
-				glog.Error(log(
-					"mounter.SetupAt failed to unpublish volume [%s]: %v (caused by previous NodePublish error: %v)",
-					c.volumeID, unpubErr, err,
-				))
-				return fmt.Errorf("%v (caused by %v)", unpubErr, err)
-			}
-
-			if unmountErr := removeMountDir(c.plugin, dir); unmountErr != nil {
-				glog.Error(log(
-					"mounter.SetupAt failed to clean mount dir [%s]: %v (caused by previous NodePublish error: %v)",
-					dir, unmountErr, err,
-				))
-				return fmt.Errorf("%v (caused by %v)", unmountErr, err)
-			}
+	// The following logic is derived from https://github.com/kubernetes/kubernetes/issues/66323
+	// if fstype is "", then skip fsgroup (could be indication of non-block filesystem)
+	// if fstype is provided and pv.AccessMode == ReadWriteOnly, then apply fsgroup
 
-			return err
+	err = applyFSGroup(c, fsType, fsGroup)
```

Comment:
Did we intentionally drop `c.readOnly` check from here?

Reply:
@vladimirvivien ping on this.

Reply:
@gnufied apologies on review.  Yeah I think I removed it because it overlaps the variables used to determine when to apply fsgroup. It looks like attribute `ReadWriteOnce` is the determining factor because you can have a `readOnceMany` that is ReadOnly then you still cant safely apply fsgroup.

Reply:
hmm, I think that is very different from how other volume plugins handles this. If nothing else - I think it should be uniform between plugins -  https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/aws_ebs/aws_ebs.go#L423

I think, current/existing behaviour is also subtly incorrect. You can specify `ReadOnly` in two places. In `pod.Spec.Volumes` and in `pod.Spec.Containers.VolumeMounts`. Currently, we seem to respect `ReadOnly` value inside only `pod.Spec.Volumes` in all the plugins and value inside `pod.Spec.Containers.VolumeMounts` does nothing.

Reply:
Ok I will take into account the readOnly in addition to the other attributes.  Keep in mind, one of the issue that was brought up (here https://github.com/kubernetes/kubernetes/issues/66323), if volattrib is other than `ReadOnlyOnce` and fsgroup is applied, it can cause issues.  So, readOnly cannot be the only determining factor.